### ISSUE TITLE
nix3-build should release memory acquired for evaluation before building the derivation

### ISSUE THREAD

**Issue by poscat0x04** [NixOS/nix#10862]
**Describe the bug**
I'm not sure if this counts as a bug report or feature request.

Basically currently `nix build` won't release the pages allocated for evaluation while waiting for `nix-daemon` to build the final derivation. Since evaluation can take a lot of RAM (easily on the order of a few gigabytes when evaluation a system configuration), this can cause problems if the host is memory starved an the build task is demanding.

The same issue can be mitigated by first calling `nix path-info --derivation` to build a `.drv` file (which does the evaluation part) and then calling `nix-store --realise` on the resulting `.drv` file (which does the build part). I feel like this counts as major space leak.

One way I can think of to solve this issue is to spawn a subprocess specifically for evaluation, this way the memory can be easily reclaimed by simply killing the subprocess. Not sure how hard it would be to implement though.

**Expected behavior**

`nix build` should release unused RAM before calling `nix-daemon`

**`nix-env --version` output**
nix-env (Nix) 2.18.2

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

**Comment by roberth** [comment]
Somewhat of a duplicate, but for the new CLI.
See https://github.com/NixOS/nix/issues/5200 for previous discussion including possible solutions and trade offs.

**Comment by poscat0x04** [comment]
Ugh sorry I couldn't find that issue :/ Should I close mine as a duplicate?

**Comment by poscat0x04** [comment]
Also my hypothesis that while the heap allocation used during evaluation got `free()`d afterwards, no calls to sbrk are made to unmap the allocated memory pages, causing them to still appear as being used to the OS. I don't have any experience with c++ or Boehm GC though so this is just a guess.

**Comment by roberth** [comment]
You might be onto something. It doesn't seem eager to unmap when reading the code docs, although the default setting is unclear.
Could you try with `GC_FORCE_UNMAP_ON_GCOLLECT=1` in the environment?
Source: https://github.com/ivmai/bdwgc/blob/8915b63d987f6034fa53d42c3a2d8a2b978f0614/include/gc/gc.h#L2150

Also if we want to run a GC before building, we should probably call [`GC_gcollect_and_unmap`](https://github.com/ivmai/bdwgc/blob/8915b63d987f6034fa53d42c3a2d8a2b978f0614/include/gc/gc.h#L761) instead of `GC_gcollect`, to basically force the same effect, temporarily.

**Comment by poscat0x04** [comment]
Yes it made a significant difference. Setting the flag makes the memory usage drop from 3.5 gigs to ~800MB.

**Comment by roberth** [comment]
@poscat0x04 could you try https://github.com/NixOS/nix/pull/10879?

**Comment by domenkozar** [comment]
https://github.com/NixOS/nix/pull/16503 or https://github.com/NixOS/nix/pull/16114 would fix it

**Comment by xokdvium** [comment]
> https://github.com/NixOS/nix/pull/16503 or https://github.com/NixOS/nix/pull/16114 would fix it

Not really though. We can't release all the roots because the EvalState itself is alive and can't be destroyed safely (or even at all, depending on where exactly refcounts to it have snuck in).

Sure, that fix is useful for the in-process C API but realistically the only way to release *everything* now is to throw away the whole process.

Also, I'm a bit skeptical that file eval cache is the only place that keeps around live closure - values produced in nixpkgs are inherently leaky due to overriding closures that keep mostly everything alive. And we can't safely drop the top-level values in most places.

**Comment by domenkozar** [comment]
> > [#16503](https://github.com/NixOS/nix/pull/16503) or [#16114](https://github.com/NixOS/nix/pull/16114) would fix it
> 
> Not really though. We can't release all the roots because the EvalState itself is alive and can't be destroyed safely (or even at all, depending on where exactly refcounts to it have snuck in).
> 
> Sure, that fix is useful for the in-process C API but realistically the only way to release _everything_ now is to throw away the whole process.
> 
> Also, I'm a bit skeptical that file eval cache is the only place that keeps around live closure - values produced in nixpkgs are inherently leaky due to overriding closures that keep mostly everything alive. And we can't safely drop the top-level values in most places.

At least majority of memory is now reclaimed for `devenv mcp`, which does full package eval.
